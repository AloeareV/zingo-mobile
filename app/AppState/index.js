import Address from './Address';
import AddressBookEntry from './AddressBookEntry';
import AppStateLoaded from './AppStateLoaded';
import AppStateLoading from './AppStateLoading';
import ConnectedCompanionApp from './ConnectedCompanionApp';
import ErrorModalData from './ErrorModalData';
import InfoType from './InfoType';
import PasswordState from './PasswordState';
import ReceivePageState from './ReceivePageState';
import SendJsonToType from './SendJsonToType';
import SendPageState from './SendPageState';
import SendProgress from './SendProgress';
import ServerSelectState from './ServerSelectState';
import SettingsFileEntry from './SettingsFileEntry';
import SyncStatus from './SyncStatus';
import SyncStatusReport from './SyncStatusReport';
import ToAddr from './ToAddr';
import TotalBalance from './TotalBalance';
import Transaction from './Transaction';
import TxDetailType from './TxDetailType';
import WalletSeed from './WalletSeed';
import WalletSettings from './WalletSettings';

export {
  AppStateLoaded,
  AppStateLoading,
  Address,
  AddressBookEntry,
  ConnectedCompanionApp,
  ErrorModalData,
  InfoType,
  PasswordState,
  ReceivePageState,
  SendJsonToType,
  SendPageState,
  SendProgress,
  ServerSelectState,
  SettingsFileEntry,
  SyncStatus,
  SyncStatusReport,
  ToAddr,
  TotalBalance,
  Transaction,
  TxDetailType,
  WalletSeed,
  WalletSettings,
};
