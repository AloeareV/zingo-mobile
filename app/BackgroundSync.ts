import RPCModule from './RPCModule';
import AsyncStorage from '@react-native-async-storage/async-storage';
import NetInfo, { NetInfoStateType } from '@react-native-community/netinfo';
import { RPCSyncStatusType } from './rpc/types/RPCSyncStatusType';

const BackgroundSync = async (task_data: any) => {
  // this not interact with the server.
  const exists = await RPCModule.walletExists();

  // only if exists the wallet file make sense to do the sync.
  if (exists && exists !== 'false') {
    // only if we have connection make sense to call RPCModule.
    const networkState = await NetInfo.fetch();
    if (
      !networkState.isConnected ||
      networkState.type === NetInfoStateType.cellular ||
      (networkState.details !== null && networkState.details.isConnectionExpensive)
    ) {
      //console.log(
      //  'BS: Not started (connected: ' + networkState.isConnected,
      //  +', type: ' +
      //    networkState.type +
      //    +', expensive connection: ' +
      //    networkState.details?.isConnectionExpensive +
      //    ')',
      //);
      return;
    }
    // if the App goes to Foreground kill the interval
    const background = await AsyncStorage.getItem('@background');
    if (background === 'no') {
      //console.log('BS: Not started (going to foreground)');
      return;
    }

    let batch_num = -1;
    console.log('BS:', task_data);

    // finishEarly has two fields: wait, and done.
    // wait() returns a promise, which is resolved when
    // done() is called
    let finishEarly = manuallyResolve();

    let saver = setInterval(async () => {
      const networkStateSaver = await NetInfo.fetch();
      if (
        !networkStateSaver.isConnected ||
        networkStateSaver.type === NetInfoStateType.cellular ||
        (networkStateSaver.details !== null && networkStateSaver.details.isConnectionExpensive)
      ) {
        //console.log(
        //  'BS: Interrupted (connected: ' + networkStateSaver.isConnected,
        //  +', type: ' +
        //    networkStateSaver.type +
        //    +', expensive connection: ' +
        //    networkStateSaver.details?.isConnectionExpensive +
        //    ')',
        //);
        clearInterval(saver);
        finishEarly.done();
        return;
      }
      // if the App goes to Foreground kill the interval
      const backgroundSaver = await AsyncStorage.getItem('@background');
      if (backgroundSaver === 'no') {
        clearInterval(saver);
        //console.log('BS: Finished (going to foreground)');
        finishEarly.done();
        return;
      }

      const syncStatusStr: string = await RPCModule.execute('syncstatus', '');
      if (syncStatusStr) {
        if (syncStatusStr.toLowerCase().startsWith('error')) {
          //console.log(`BS: Error sync status ${syncStatusStr}`);
          return;
        }
      } else {
        //console.log('BS: Internal Error sync status');
        return;
      }
      // TODO: verify this JSON parse
      const ss: RPCSyncStatusType = await JSON.parse(syncStatusStr);

      //console.log('BS:', ss);
      if (ss.batch_num && ss.batch_num > -1 && batch_num !== ss.batch_num) {
        await RPCModule.doSave();
        //console.log('BS: saving...');
        // update batch_num with the new value, otherwise never change
        batch_num = ss.batch_num;
      }
    }, 5000);

    await Promise.race([RPCModule.execute('sync', ''), finishEarly.wait()]);
    clearInterval(saver);
  } else {
    console.log('BS: wallet file does not exist');
  }
  //console.log('BS: Finished (end of syncing)');
};

export default BackgroundSync;

function manuallyResolve() {
  let resolve: Function;
  // new Promise takes a function as an argument. When that function is called
  // the promise resolves with the value output by that function.
  // By passing the function out of the promise, we can call it later
  // in order to resolve the promise at will
  const promise = new Promise(fun => {
    resolve = fun;
  });

  function done() {
    resolve();
  }

  function wait() {
    return promise;
  }

  return { wait, done };
}
