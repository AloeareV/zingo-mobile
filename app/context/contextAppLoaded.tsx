import React, { ReactNode } from 'react';
import { StackScreenProps } from '@react-navigation/stack';

import {
  AppStateLoaded,
  SyncingStatusReportClass,
  TotalBalanceClass,
  ReceivePageStateClass,
  ErrorModalDataClass,
  SendProgressClass,
  WalletSettingsClass,
  TransactionType,
  InfoType,
  SyncingStatusType,
  WalletSeedType,
  DimensionsType,
  AddressClass,
  AddressBookClass,
  zecPriceType,
  BackgroundType,
  SendPageStateClass,
  ToAddrClass,
  NetInfoType,
  BackgroundErrorType,
} from '../AppState';

export const defaultAppStateLoaded: AppStateLoaded = {
  navigation: {} as StackScreenProps<any>['navigation'],
  route: {} as StackScreenProps<any>['route'],
  dimensions: {} as DimensionsType,
  appState: '' as string,
  netInfo: {} as NetInfoType,

  syncingStatusReport: new SyncingStatusReportClass(),
  totalBalance: new TotalBalanceClass(),
  addressPrivateKeys: new Map(),
  addresses: [] as AddressClass[],
  addressBook: [] as AddressBookClass[],
  transactions: [] as TransactionType[],
  sendPageState: new SendPageStateClass(new ToAddrClass(0)),
  receivePageState: new ReceivePageStateClass(''),
  info: {} as InfoType,
  walletSettings: new WalletSettingsClass(),
  syncingStatus: {} as SyncingStatusType,
  errorModalData: new ErrorModalDataClass('', ''),
  sendProgress: new SendProgressClass(0, 0, 0),
  walletSeed: {} as WalletSeedType,
  isMenuDrawerOpen: false,
  selectedMenuDrawerItem: '' as string,
  aboutModalVisible: false,
  computingModalVisible: false,
  settingsModalVisible: false,
  infoModalVisible: false,
  rescanModalVisible: false,
  seedViewModalVisible: false,
  seedChangeModalVisible: false,
  seedBackupModalVisible: false,
  seedServerModalVisible: false,
  syncReportModalVisible: false,
  poolsModalVisible: false,
  newServer: '' as string,
  uaAddress: '' as string,

  server: '' as string,
  currency: '',
  language: 'en',

  zecPrice: {
    zecPrice: 0,
    date: 0,
  } as zecPriceType,
  sendAll: false,
  background: {
    batches: 0,
    date: 0,
  } as BackgroundType,

  translate: () => '',
  backgroundError: {} as BackgroundErrorType,
};

export const ContextAppLoaded = React.createContext(defaultAppStateLoaded);

type ContextProviderProps = {
  children: ReactNode;
  value: AppStateLoaded;
};

export const ContextAppLoadedProvider = ({ children, value }: ContextProviderProps) => {
  return <ContextAppLoaded.Provider value={value}>{children}</ContextAppLoaded.Provider>;
};
