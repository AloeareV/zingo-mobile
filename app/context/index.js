import { defaultAppStateLoaded, ContextAppLoaded, ContextAppLoadedProvider } from './contextAppLoaded';
import { defaultAppStateLoading, ContextAppLoading, ContextAppLoadingProvider } from './contextAppLoading';

export {
  defaultAppStateLoaded,
  ContextAppLoaded,
  ContextAppLoadedProvider,
  defaultAppStateLoading,
  ContextAppLoading,
  ContextAppLoadingProvider,
};
