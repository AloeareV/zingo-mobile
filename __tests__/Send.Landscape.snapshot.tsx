/**
 * @format
 */

import 'react-native';
import React from 'react';
import { StackScreenProps } from '@react-navigation/stack';

import { render } from '@testing-library/react-native';
import Send from '../components/Send';
import { ContextLoadedProvider } from '../app/context';

import {
  Address,
  AddressBookEntry,
  ErrorModalData,
  InfoType,
  ReceivePageState,
  SendPageState,
  SendProgress,
  SyncStatus,
  SyncStatusReport,
  ToAddr,
  TotalBalance,
  Transaction,
  WalletSeed,
  WalletSettings,
} from '../app/AppState';

jest.useFakeTimers();
jest.mock('@fortawesome/react-native-fontawesome', () => ({
  FontAwesomeIcon: '',
}));
jest.mock('react-native-localize', () => ({
  getNumberFormatSettings: () => {
    return {
      decimalSeparator: '.', // us
      groupingSeparator: ',', // us
    };
  },
}));
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
jest.mock('react-native-reanimated', () => {
  return class Reanimated {
    public static Value() {
      return jest.fn(() => {});
    }
    public static View() {
      return '';
    }
  };
});

// test suite
describe('Component Send - test', () => {
  //snapshot test
  test('Send Landscape - snapshot', () => {
    const state = {
      navigation: {} as StackScreenProps<any>['navigation'],
      route: {} as StackScreenProps<any>['route'],

      syncStatusReport: new SyncStatusReport(),
      addressPrivateKeys: new Map(),
      addressBook: [] as AddressBookEntry[],
      transactions: [
        {
          type: 'sent',
          address: 'sent-address-12345678901234567890',
          amount: 0.12345678,
          position: '',
          confirmations: 22,
          txid: 'sent-txid-1234567890',
          time: Date.now(),
          zec_price: 33.33,
          detailedTxns: [],
        },
        {
          type: 'receive',
          address: 'receive-address-12345678901234567890',
          amount: 0.87654321,
          position: '',
          confirmations: 133,
          txid: 'receive-txid-1234567890',
          time: Date.now(),
          zec_price: 66.66,
          detailedTxns: [],
        },
      ] as Transaction[],
      sendPageState: new SendPageState(new ToAddr(0)),
      receivePageState: new ReceivePageState(),
      info: {} as InfoType,
      rescanning: false,
      wallet_settings: new WalletSettings(),
      syncingStatus: {} as SyncStatus,
      errorModalData: new ErrorModalData(),
      sendProgress: new SendProgress(),
      walletSeed: {} as WalletSeed,
      isMenuDrawerOpen: false,
      selectedMenuDrawerItem: '' as string,
      aboutModalVisible: false,
      computingModalVisible: false,
      settingsModalVisible: false,
      infoModalVisible: false,
      rescanModalVisible: false,
      seedViewModalVisible: false,
      seedChangeModalVisible: false,
      seedBackupModalVisible: false,
      seedServerModalVisible: false,
      syncReportModalVisible: false,
      poolsModalVisible: false,
      newServer: '' as string,
      uaAddress: 'UA-12345678901234567890',
      addresses: [
        {
          uaAddress: 'UA-12345678901234567890',
          address: 'UA-12345678901234567890',
          addressKind: 'u',
          containsPending: false,
          receivers: 'ozt',
        },
        {
          uaAddress: 'UA-12345678901234567890',
          address: 'sapling-12345678901234567890',
          addressKind: 'z',
          containsPending: false,
          receivers: 'z',
        },
        {
          uaAddress: 'UA-12345678901234567890',
          address: 'transparent-12345678901234567890',
          addressKind: 't',
          containsPending: false,
          receivers: 't',
        },
      ] as Address[],
      translate: () => 'text translated',
      dimensions: {
        width: 600,
        height: 300,
        orientation: 'landscape',
        deviceType: 'phone',
        scale: 2.5,
      } as {
        width: number;
        height: number;
        orientation: 'portrait' | 'landscape';
        deviceType: 'tablet' | 'phone';
        scale: number;
      },
      totalBalance: new TotalBalance(),
    };
    state.info.currencyName = 'ZEC';
    state.info.zecPrice = 33.33;
    state.info.defaultFee = 1000;
    state.totalBalance.total = 1.12345678;
    state.totalBalance.orchardBal = 0.6;
    state.totalBalance.spendableOrchard = 0.3;
    state.totalBalance.privateBal = 0.4;
    state.totalBalance.spendablePrivate = 0.2;
    state.totalBalance.transparentBal = 0.12345678;
    state.sendPageState.toaddr.id = 1234567890;
    state.sendPageState.toaddr.to = 'UA-12345678901234567890';
    state.sendPageState.toaddr.amount = '1.12345678';
    state.sendPageState.toaddr.amountUSD = '50.22';
    state.sendPageState.toaddr.memo = 'memo';
    const onFunction = jest.fn();
    const send = render(
      <ContextLoadedProvider value={state}>
        <Send
          setSendPageState={onFunction}
          sendTransaction={onFunction}
          clearToAddr={onFunction}
          setSendProgress={onFunction}
          toggleMenuDrawer={onFunction}
          setComputingModalVisible={onFunction}
          poolsMoreInfoOnClick={onFunction}
          syncingStatusMoreInfoOnClick={onFunction}
        />
      </ContextLoadedProvider>,
    );
    expect(send.toJSON()).toMatchSnapshot();
  });
});
