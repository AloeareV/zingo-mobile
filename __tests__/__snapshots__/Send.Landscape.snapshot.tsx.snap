// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Send - test Send Landscape - snapshot 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "height": "100%",
    }
  }
>
  <View
    accessibilityLabel="text translated"
    accessible={true}
    style={
      {
        "display": "flex",
        "justifyContent": "flex-start",
        "width": 300,
      }
    }
  >
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={false}
      visible={false}
    />
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={false}
      visible={false}
    />
    <View
      style={{}}
    >
      <View
        onLayout={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgb(255, 255, 255)",
            "display": "flex",
            "zIndex": -1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgb(255, 255, 255)",
              "display": "flex",
              "padding": 10,
              "width": "100%",
              "zIndex": -1,
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../assets/img/logobig-zingo.png",
              }
            }
            style={
              {
                "height": 80,
                "resizeMode": "contain",
                "width": 80,
              }
            }
          />
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "column",
              }
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "flex-end",
                    "backgroundColor": "rgb(255, 255, 255)",
                    "borderRadius": 10,
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "margin": 0,
                    "minHeight": 48,
                    "minWidth": 48,
                    "padding": 0,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "rgb(0, 122, 255)",
                        "fontSize": 18,
                        "fontWeight": "600",
                        "opacity": 1,
                      },
                    ]
                  }
                >
                  text translated
                </Text>
                <
                  color="rgb(0, 122, 255)"
                  icon={
                    {
                      "icon": [
                        192,
                        512,
                        [],
                        "f129",
                        "M144 80c0 26.5-21.5 48-48 48s-48-21.5-48-48s21.5-48 48-48s48 21.5 48 48zM0 224c0-17.7 14.3-32 32-32H96c17.7 0 32 14.3 32 32V448h32c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H64V256H32c-17.7 0-32-14.3-32-32z",
                      ],
                      "iconName": "info",
                      "prefix": "fas",
                    }
                  }
                  size={14}
                  style={
                    {
                      "marginBottom": 5,
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "baseline",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  {
                    "color": undefined,
                    "fontSize": 36,
                  }
                }
              >
                á™‡
              </Text>
              <Text
                style={
                  {
                    "color": undefined,
                    "fontSize": 36,
                    "fontWeight": "700",
                  }
                }
              >
                 1.1234
              </Text>
              <Text
                style={
                  {
                    "color": undefined,
                    "fontSize": 18,
                    "paddingBottom": 0,
                  }
                }
              >
                5678
              </Text>
              <Text
                style={
                  {
                    "color": undefined,
                    "fontSize": 36,
                  }
                }
              >
                 ZEC
              </Text>
            </View>
            <View
              style={
                {
                  "alignItems": "baseline",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  {
                    "color": undefined,
                    "fontSize": 20,
                    "marginBottom": 5,
                    "marginTop": 0,
                  }
                }
              >
                $
              </Text>
              <Text
                style={
                  {
                    "color": undefined,
                    "fontSize": 20,
                    "fontWeight": "700",
                    "marginBottom": 5,
                    "marginTop": 0,
                  }
                }
              >
                 37.44
              </Text>
              <Text
                style={
                  {
                    "color": undefined,
                    "fontSize": 20,
                    "marginBottom": 5,
                    "marginTop": 0,
                  }
                }
              >
                 USD
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "backgroundColor": "rgb(0, 122, 255)",
                "height": 1,
                "marginTop": 5,
                "width": "100%",
              }
            }
          />
          <View
            style={
              {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "flexWrap": "wrap",
                "justifyContent": "center",
                "marginVertical": 5,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "column",
                  "flexWrap": "wrap",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "paddingHorizontal": 5,
                    },
                    {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 18,
                      "fontWeight": "600",
                      "opacity": 1,
                    },
                  ]
                }
              >
                text translated
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "backgroundColor": "rgb(0, 122, 255)",
                "height": 1,
                "width": "100%",
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "rgb(255, 255, 255)",
          "padding": 10,
          "position": "absolute",
        }
      }
    >
      <View
        accessibilityLabel="text translated"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <
          color="rgb(216, 216, 216)"
          icon={
            {
              "icon": [
                448,
                512,
                [
                  "navicon",
                ],
                "f0c9",
                "M0 96C0 78.3 14.3 64 32 64H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H416c17.7 0 32 14.3 32 32z",
              ],
              "iconName": "bars",
              "prefix": "fas",
            }
          }
          size={48}
        />
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "borderLeftColor": "rgb(216, 216, 216)",
        "borderLeftWidth": 1,
        "height": "100%",
        "padding": 0,
        "width": 300,
      }
    }
  >
    <RCTScrollView
      contentContainerStyle={{}}
      keyboardShouldPersistTaps="handled"
      style={{}}
    >
      <View>
        <View
          style={
            {
              "marginBottom": 150,
            }
          }
        >
          <View
            style={
              {
                "display": "flex",
                "marginTop": 10,
                "padding": 5,
              }
            }
          >
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 18,
                      "fontWeight": "600",
                      "opacity": 1,
                    },
                  ]
                }
              >
                text translated
              </Text>
            </View>
            <View
              style={
                {
                  "borderColor": "rgb(28, 28, 30)",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "flex": 1,
                  "marginTop": 5,
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  accessibilityLabel="text translated"
                  accessible={true}
                  style={
                    {
                      "flex": 1,
                      "justifyContent": "center",
                    }
                  }
                >
                  <TextInput
                    editable={true}
                    onChangeText={[Function]}
                    placeholder="text translated"
                    style={
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 16,
                        "fontWeight": "600",
                        "marginLeft": 5,
                      }
                    }
                    value="UA-12345678901234567890"
                  />
                </View>
                <View
                  style={
                    {
                      "width": 58,
                    }
                  }
                >
                  <View
                    accessibilityLabel="text translated"
                    accessible={true}
                    collapsable={false}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      {
                        "opacity": 1,
                      }
                    }
                  >
                    <
                      color="rgb(216, 216, 216)"
                      icon={
                        {
                          "icon": [
                            448,
                            512,
                            [],
                            "f029",
                            "M48 32C21.5 32 0 53.5 0 80v96c0 26.5 21.5 48 48 48h96c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48H48zm80 64v64H64V96h64zM48 288c-26.5 0-48 21.5-48 48v96c0 26.5 21.5 48 48 48h96c26.5 0 48-21.5 48-48V336c0-26.5-21.5-48-48-48H48zm80 64v64H64V352h64zM256 80v96c0 26.5 21.5 48 48 48h96c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48H304c-26.5 0-48 21.5-48 48zm64 16h64v64H320V96zm32 352v32h32V448H352zm96 0H416v32h32V448zM416 288v32H352V288H256v96 96h64V384h32v32h96V352 320 288H416z",
                          ],
                          "iconName": "qrcode",
                          "prefix": "fas",
                        }
                      }
                      size={48}
                      style={
                        {
                          "margin": 5,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginTop": 10,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "rgb(28, 28, 30)",
                    "opacity": 0.65,
                  }
                }
              >
                text translated
              </Text>
              <Text
                style={
                  {
                    "color": "rgb(0, 122, 255)",
                  }
                }
              >
                text translated
              </Text>
            </View>
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <View
                style={
                  {
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "width": "60%",
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontSize": 20,
                        "marginRight": 5,
                        "marginTop": 20,
                      },
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 18,
                        "fontWeight": "600",
                        "opacity": 1,
                      },
                    ]
                  }
                >
                  á™‡
                </Text>
                <View
                  accessibilityLabel="text translated"
                  accessible={true}
                  style={
                    {
                      "borderColor": "rgb(28, 28, 30)",
                      "borderRadius": 5,
                      "borderWidth": 1,
                      "flexGrow": 1,
                      "marginTop": 5,
                      "minHeight": 48,
                      "minWidth": 48,
                      "width": "75%",
                    }
                  }
                >
                  <TextInput
                    editable={true}
                    keyboardType="numeric"
                    maxLength={20}
                    onChangeText={[Function]}
                    onEndEditing={[Function]}
                    placeholder="#.########"
                    style={
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 18,
                        "fontWeight": "600",
                        "marginLeft": 5,
                        "minHeight": 48,
                        "minWidth": 48,
                      }
                    }
                    value="1.12345678"
                  />
                </View>
                <Text
                  style={
                    [
                      {
                        "marginLeft": 5,
                        "marginRight": 10,
                        "marginTop": 15,
                      },
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 18,
                        "fontWeight": "600",
                        "opacity": 1,
                      },
                    ]
                  }
                >
                  ZEC
                </Text>
              </View>
              <View
                style={
                  {
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "width": "35%",
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "marginRight": 5,
                        "marginTop": 15,
                      },
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 18,
                        "fontWeight": "600",
                        "opacity": 1,
                      },
                    ]
                  }
                >
                  $
                </Text>
                <View
                  accessibilityLabel="text translated"
                  accessible={true}
                  style={
                    {
                      "borderColor": "rgb(28, 28, 30)",
                      "borderRadius": 5,
                      "borderWidth": 1,
                      "flexGrow": 1,
                      "marginTop": 5,
                      "minHeight": 48,
                      "minWidth": 48,
                      "width": "55%",
                    }
                  }
                >
                  <TextInput
                    editable={true}
                    keyboardType="numeric"
                    maxLength={15}
                    onChangeText={[Function]}
                    onEndEditing={[Function]}
                    placeholder="#.##"
                    style={
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 18,
                        "fontWeight": "600",
                        "marginLeft": 5,
                        "minHeight": 48,
                        "minWidth": 48,
                      }
                    }
                    value="50.22"
                  />
                </View>
                <Text
                  style={
                    [
                      {
                        "marginLeft": 5,
                        "marginTop": 15,
                      },
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 18,
                        "fontWeight": "600",
                        "opacity": 1,
                      },
                    ]
                  }
                >
                  USD
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "marginTop": 10,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 18,
                        "fontWeight": "600",
                        "opacity": 1,
                      },
                    ]
                  }
                >
                  text translated
                </Text>
                <View
                  style={
                    {
                      "alignItems": "baseline",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": undefined,
                        "fontSize": 18,
                      }
                    }
                  >
                    á™‡
                  </Text>
                  <Text
                    style={
                      {
                        "color": undefined,
                        "fontSize": 18,
                        "fontWeight": "700",
                      }
                    }
                  >
                     0.0000
                  </Text>
                  <Text
                    style={
                      {
                        "color": undefined,
                        "fontSize": 9,
                        "paddingBottom": 0,
                      }
                    }
                  >
                    0000
                  </Text>
                  <Text
                    style={
                      {
                        "color": undefined,
                        "fontSize": 18,
                      }
                    }
                  >
                     ZEC
                  </Text>
                </View>
              </View>
              <View
                style={
                  {
                    "backgroundColor": "rgb(255, 255, 255)",
                    "borderRadius": 10,
                    "display": "flex",
                    "flexDirection": "row",
                    "marginTop": 5,
                    "padding": 5,
                  }
                }
              >
                <
                  color="rgb(0, 122, 255)"
                  icon={
                    {
                      "icon": [
                        192,
                        512,
                        [],
                        "f129",
                        "M144 80c0 26.5-21.5 48-48 48s-48-21.5-48-48s21.5-48 48-48s48 21.5 48 48zM0 224c0-17.7 14.3-32 32-32H96c17.7 0 32 14.3 32 32V448h32c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H64V256H32c-17.7 0-32-14.3-32-32z",
                      ],
                      "iconName": "info",
                      "prefix": "fas",
                    }
                  }
                  size={14}
                />
                <Text
                  style={
                    {
                      "color": "rgb(28, 28, 30)",
                      "opacity": 0.65,
                    }
                  }
                >
                  text translated
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "flexGrow": 1,
                "justifyContent": "center",
                "marginVertical": 5,
              }
            }
          >
            <View
              accessibilityLabel="title text translated"
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": undefined,
                  "borderColor": "rgb(0, 122, 255)",
                  "borderRadius": 10,
                  "borderWidth": 2,
                  "justifyContent": "center",
                  "maxWidth": "90%",
                  "minHeight": 48,
                  "minWidth": "30%",
                  "opacity": 1,
                  "padding": 0,
                  "paddingLeft": 20,
                  "paddingRight": 20,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "margin": 0,
                    "minHeight": 48,
                    "minWidth": 48,
                    "padding": 0,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "rgb(242, 242, 242)",
                      "fontSize": 16,
                      "fontWeight": "bold",
                      "textTransform": "uppercase",
                    }
                  }
                >
                  text translated
                </Text>
              </View>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderColor": "rgb(0, 122, 255)",
                  "borderRadius": 10,
                  "borderWidth": 2,
                  "justifyContent": "center",
                  "marginLeft": 10,
                  "maxWidth": "90%",
                  "minHeight": 48,
                  "minWidth": "30%",
                  "opacity": 1,
                  "padding": 0,
                  "paddingLeft": 20,
                  "paddingRight": 20,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "margin": 0,
                    "minHeight": 48,
                    "minWidth": 48,
                    "padding": 0,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "rgb(0, 122, 255)",
                      "fontSize": 16,
                      "fontWeight": "bold",
                      "textTransform": "uppercase",
                    }
                  }
                >
                  text translated
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
